### PR TITLE
[SPARK-38284][SQL] Add the `TIMESTAMPDIFF()` function

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to add new function `TIMESTAMPDIFF` with the following parameters:
1. `unit` - sthis indicates the units of the difference between the given timestamps. Supported string values of `unit` are (case insensitive):
   - YEAR
   - QUARTER
   - MONTH
   - WEEK
   - DAY
   - HOUR
   - MINUTE
   - SECOND
   - MILLISECOND
   - MICROSECOND
2. `timestampStart` - A timestamp which the expression subtracts from `timestampEnd`. Supported types are `timestamp` (`timestamp_ltz`).  The `timestamp_ntz` and `date` values are casted to `timestamp_ltz` implicitly at the session time zone (see the SQL config `spark.sql.session.timeZone`).
3. `timestampEnd` - A timestamp from which the expression subtracts `timestampStart`.  Supported types are `timestamp` (`timestamp_ltz`). The `timestamp_ntz` and `date` values are casted to `timestamp_ltz` implicitly at the session time zone.

The function calculates the amount in terms of this unit, and returns a whole number, representing the number of complete units between the two timestamps. The type of the result is `LONG`. The result will be negative if the end is before the start timestamp.

For example:
```sql
scala> val df = sql("select timestampdiff(YEAR, timestamp'2010-02-16 01:02:03', timestamp'2022-02-16 01:02:03') as `diff in years`")
df: org.apache.spark.sql.DataFrame = [diff in years: bigint]

scala> df.printSchema
root
 |-- diff in years: long (nullable = false)


scala> df.show(false)
+-------------+
|diff in years|
+-------------+
|12           |
+-------------+
```

### Why are the changes needed?
1. To make the migration process from other systems to Spark SQL easier.
2. To achieve feature parity with other DBMSs.

### Does this PR introduce _any_ user-facing change?
No. This is new feature.

### How was this patch tested?
By running new tests:
```
$ build/sbt "test:testOnly *QueryExecutionErrorsSuite"
$ build/sbt "test:testOnly *DateTimeUtilsSuite"
$ build/sbt "sql/test:testOnly org.apache.spark.sql.expressions.ExpressionInfoSuite"
$ build/sbt "sql/testOnly *ExpressionsSchemaSuite"
$ build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z timestamp.sql"
$ build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z timestamp-ansi.sql"
$ build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z datetime-legacy.sql"
$ build/sbt "test:testOnly *DateExpressionsSuite"
$ build/sbt "test:testOnly *SQLKeywordSuite"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@HyukjinKwon Is this known issue? One GA is failing with this but I do believe it is not related to my changes.
```
annotations failed mypy checks:
python/pyspark/pandas/generic.py:585: error: Incompatible return value type (got "Union[ndarray[Any, Any], ExtensionArray]", expected "ndarray[Any, Any]")  [return-value]
Found 1 error in 1 file (checked 3[24](https://github.com/MaxGekk/spark/runs/5294084279?check_suite_focus=true#step:15:24) source files)
1
Error: Process completed with exit code 1.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 1]
```diff
@@ -3141,3 +3141,88 @@ case class TimestampAdd(
     copy(unit = newFirst, quantity = newSecond, timestamp = newThird)
   }
 }
+
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(unit, startTimestamp, endTimestamp) - Gets the difference between the timestamps `endTimestamp` and `startTimestamp` in the specified units.",
+  arguments = """
+    Arguments:
+      * unit - this indicates the units of the difference between the given timestamps.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:3150]: shall we define the rounding behavior? e.g. if the actual result is 1.6 year, what do we return?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:3150]: The behavior has been described in PR's description already:
"... The function calculates the amount in terms of this unit, and returns a **whole** number, representing the number of **complete units between the two timestamps**. "

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:3150]: end-users can't see PR descriptions... We need to put them in the function doc.

BTW what does "whole number" mean? 1.6 years will become 2 years?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:3150]: whole and complete means 1 year for 1.6 years.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:3150]: ok so it simply truncates the fraction part. Can we clarify it in the function doc?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala [hunk 2]
```diff
@@ -989,4 +989,58 @@ class DateTimeUtilsSuite extends SparkFunSuite with Matchers with SQLHelper {
...
+        date(2022, 2, 14, 11, 0, 1, 0, zid),
+        date(2022, 2, 15, 11, 0, 1, 0, zid),
+        zid) === 24)
+      assert(timestampDiff(
+        "DAY",
+        date(2022, 2, 28, 11, 1, 0, 0, zid),
+        date(2022, 3, 1, 11, 1, 0, 0, zid),
+        zid) === 1)
+      assert(timestampDiff("WEEK",
+        date(2022, 2, 14, 11, 43, 0, 1, zid), date(2022, 2, 21, 11, 43, 0, 1, zid), zid) === 1)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala:1020]: can we test more corner cases? e.g. `date(2022, 2, 14, 11, 43, 0, 1, zid), date(2022, 2, 21, 11, 43, 0, 0, zid)` (one week minus 1 second)

**Comment by HyukjinKwon** [comment]
https://github.com/apache/spark/pull/35607#issuecomment-1048554196 @MaxGekk Just saw this. It's fixed in https://github.com/apache/spark/commit/b46b74ce0521d1d5e7c09cadad0e9639e31214cb

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: done. I modified some existing checks.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 3]
```diff
@@ -860,7 +860,8 @@ valueExpression
 
 primaryExpression
     : name=(CURRENT_DATE | CURRENT_TIMESTAMP | CURRENT_USER)                                   #currentLike
-    | TIMESTAMPADD '(' unit=identifier ',' unitsAmount=valueExpression ',' timestamp=valueExpression ')'  #timestampadd
+    | TIMESTAMPADD '(' unit=identifier ',' unitsAmount=valueExpression ',' timestamp=valueExpression ')'     #timestampadd
+    | TIMESTAMPDIFF '(' unit=identifier ',' startTimestamp=valueExpression ',' startEnd=valueExpression ')'  #timestampdiff
```
Comment (gengliangwang) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:864]: nit: `startEnd` => `endTimestamp`

**Review by gengliangwang — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 4]
```diff
@@ -3131,3 +3131,88 @@ case class TimestampAdd(
...
+
+  def this(unit: Expression, quantity: Expression, timestamp: Expression) =
+    this(unit, quantity, timestamp, None)
+
+  override def first: Expression = unit
+  override def second: Expression = startTimestamp
+  override def third: Expression = endTimestamp
+
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(StringType, AnyTimestampType, AnyTimestampType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:3187]: I think another requirement is the start and end timestamp must be of the same type, if one is ntz and the other is ltz, we should cast ntz to ltz. cc @gengliangwang

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:3187]: or should we just require both types to be ltz?

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:3187]: @cloud-fan good point.
I just tried running  some tests:
```
> select timestampdiff(second, timestamp_ntz'2021-01-01 00:00:00', timestamp_ltz'2021-01-01 00:00:00'
28800

> select timestamp_ntz'2021-01-01 00:00:00' - timestamp_ltz'2021-01-01 00:00:00'
INTERVAL '0 00:00:00' DAY TO SECOND  
```
I think we can simply write the input type as TimestampType and the implicit cast trait will handle the rest input combination.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:3187]: hmm, the results between `timestampdiff` and `-` operator are not consistent now?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:3187]: ok seems like a bug and if we require both inputs to be ltz the bug will be fixed.

**Review by cloud-fan — APPROVED** [review]

**Pull request closed without merging**